# Importing Necessary Libraries

In [35]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from warnings import filterwarnings

In [36]:
filterwarnings('ignore')

In [37]:
df = pd.read_csv('air quality data.csv')

# Reviewing basic Data

In [38]:
df.sample(10)

City        Date   PM2.5    PM10     NO     NO2     NOx  \
10648          Delhi  2016-02-24  177.64  309.47  93.53  117.04  150.74   
22203         Mumbai  2015-12-15     NaN     NaN    NaN     NaN   97.43   
29144  Visakhapatnam  2019-06-11   37.53  113.19   8.34   49.11   32.91   
11034          Delhi  2017-03-16  125.80  348.95  73.95   42.18   35.00   
24464          Patna  2017-01-21  140.83     NaN  28.93   49.75   78.68   
22813         Mumbai  2017-08-16     NaN     NaN    NaN     NaN     NaN   
10925          Delhi  2016-11-27  128.70  336.16  35.91   71.29   40.84   
11007          Delhi  2017-02-17  112.02  330.90  50.86   72.55   52.64   
5317       Bengaluru  2017-10-20   79.76  264.90  14.24   40.38   22.47   
23204         Mumbai  2018-09-11   26.76   72.02  51.88   23.19   74.94   

         NH3    CO    SO2      O3  Benzene  Toluene  Xylene    AQI  \
10648  35.85  2.07  50.44  200.41     4.50    14.62     NaN  319.0   
22203    NaN  0.00    NaN     NaN     0.00     0.00    0.00    NaN   
29144   3.74  0.74  14.49   24.99     5.61    13.34    3.17   99.0   
11034  43.07  0.52  23.59   31.90     3.93     9.51     NaN  371.0   
24464    NaN  1.59   3.96   22.35     0.13     0.52    0.17  301.0   
22813    NaN  0.00    NaN     NaN     0.00     0.00    0.00    NaN   
10925  26.73  0.65  22.68   38.99     3.16     5.83     NaN  347.0   
11007  53.40  0.71  19.33   32.42     3.42     6.45     NaN  309.0   
5317   26.81  0.58   5.88   38.72     1.60     2.51     NaN  177.0   
23204    NaN  1.55  13.90   22.46     0.99      NaN     NaN  106.0   

         AQI_Bucket  
10648     Very Poor  
22203           NaN  
29144  Satisfactory  
11034     Very Poor  
24464     Very Poor  
22813           NaN  
10925     Very Poor  
11007     Very Poor  
5317       Moderate  
23204      Moderate

In [39]:
df.shape

(29531, 16)

In [40]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29531 entries, 0 to 29530
Data columns (total 16 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   City        29531 non-null  object 
 1   Date        29531 non-null  object 
 2   PM2.5       24933 non-null  float64
 3   PM10        18391 non-null  float64
 4   NO          25949 non-null  float64
 5   NO2         25946 non-null  float64
 6   NOx         25346 non-null  float64
 7   NH3         19203 non-null  float64
 8   CO          27472 non-null  float64
 9   SO2         25677 non-null  float64
 10  O3          25509 non-null  float64
 11  Benzene     23908 non-null  float64
 12  Toluene     21490 non-null  float64
 13  Xylene      11422 non-null  float64
 14  AQI         24850 non-null  float64
 15  AQI_Bucket  24850 non-null  object 
dtypes: float64(13), object(3)
memory usage: 3.6+ MB


In [41]:
df.isnull().sum()

City              0
Date              0
PM2.5          4598
PM10          11140
NO             3582
NO2            3585
NOx            4185
NH3           10328
CO             2059
SO2            3854
O3             4022
Benzene        5623
Toluene        8041
Xylene        18109
AQI            4681
AQI_Bucket     4681
dtype: int64

In [42]:
df.duplicated().sum()

0

Cleaning mising AQI rows as its our target

In [23]:
df.dropna(subset=['AQI'],inplace=True)

In [24]:
df.isnull().sum()

City              0
Date              0
PM2.5           678
PM10           7086
NO              387
NO2             391
NOx            1857
NH3            6536
CO              445
SO2             605
O3              807
Benzene        3535
Toluene        5826
Xylene        15372
AQI               0
AQI_Bucket        0
dtype: int64

In [26]:
df.shape

(24850, 16)

# **Summary of Dataset**

In [28]:
df.describe().T # transpose to make col to rows andvice versa

count        mean         std    min      25%      50%       75%  \
PM2.5    24172.0   67.476613   63.075398   0.04  29.0000   48.785   80.9250   
PM10     17764.0  118.454435   89.487976   0.03  56.7775   96.180  150.1825   
NO       24463.0   17.622421   22.421138   0.03   5.6600    9.910   20.0300   
NO2      24459.0   28.978391   24.627054   0.01  11.9400   22.100   38.2400   
NOx      22993.0   32.289012   30.712855   0.00  13.1100   23.680   40.1700   
NH3      18314.0   23.848366   25.875981   0.01   8.9600   16.310   30.3600   
CO       24405.0    2.345267    7.075208   0.00   0.5900    0.930    1.4800   
SO2      24245.0   14.362933   17.428693   0.01   5.7300    9.220   15.1400   
O3       24043.0   34.912885   21.724525   0.01  19.2500   31.250   46.0800   
Benzene  21315.0    3.458668   16.036020   0.00   0.2300    1.290    3.3400   
Toluene  19024.0    9.525714   20.881085   0.00   1.0275    3.575   10.1800   
Xylene    9478.0    3.588683    6.754324   0.00   0.3900    1.420    4.1200   
AQI      24850.0  166.463581  140.696585  13.00  81.0000  118.000  208.0000   

             max  
PM2.5     914.94  
PM10      917.08  
NO        390.68  
NO2       362.21  
NOx       378.24  
NH3       352.89  
CO        175.81  
SO2       186.08  
O3        257.73  
Benzene   455.03  
Toluene   454.85  
Xylene    170.37  
AQI      2049.00

# **Percentage of Null values**

In [33]:
null_Percent = (df.isnull().sum() / df.isnull().count() * 100).sort_values(ascending=False)

In [34]:
null_Percent

Xylene        61.859155
PM10          28.515091
NH3           26.301811
Toluene       23.444668
Benzene       14.225352
NOx            7.472837
O3             3.247485
PM2.5          2.728370
SO2            2.434608
CO             1.790744
NO2            1.573441
NO             1.557344
City           0.000000
Date           0.000000
AQI            0.000000
AQI_Bucket     0.000000
dtype: float64


*   **Xylene** has highest missing values --> 61.88%
*   **Ph-10** and **NH3** --> 26%-28%